drivers/can/can_stm32.c

```diff
@@ -415,6 +419,12 @@ static int can_stm32_stop(const struct device *dev)
 		goto unlock;
 	}
 
+	/* Abort any pending transmissions */
+	can_stm32_signal_tx_complete(dev, &data->mb0, -ENETDOWN);
+	can_stm32_signal_tx_complete(dev, &data->mb1, -ENETDOWN);
+	can_stm32_signal_tx_complete(dev, &data->mb2, -ENETDOWN);
+	can->TSR |= CAN_TSR_ABRQ2 | CAN_TSR_ABRQ1 | CAN_TSR_ABRQ0;
```

Comment:
That passage in the data sheet is exactly why it needs to be in this order:
- Entering init mode will wait for the current transmission (move-out) to complete, if any is in progress (but it will not start a new move-out).
- In order for the remaining, pending frames not to be transmitted after calling `can_start()` they are aborted just after entering init mode.

Reply:
To be clear, calling `can_stop()` will abort _pending_ transmissions, not an _in-progress_ transmission. So when requesting to enter init mode, the bxCAN controller will wait for an _in-progress_ transmission to finish (either with success or failure) and then enter init mode.

Setting the abort requests is solely done in order to "flush" the bxCAN controller TX MBs.

Reply:
I don't fully understand the **Abort** section of the documentation:

> An abort request while the mailbox is in transmit state can have two results. If the mailbox is transmitted successfully the mailbox becomes empty with the TXOK bit set in the CAN_TSR register. If the transmission fails, the mailbox becomes scheduled, the transmission is aborted and becomes empty with TXOK cleared.

Does this mean that the abort request would immediately stop the transmission of a frame even in the middle of a frame? If that's the case, it would be pretty bad behavior.

Looks like the current approach is safe. Init mode stops further transmissions and the queues are cleared afterwards.

Reply:
> Does this mean that the abort request would immediately stop the transmission of a frame even in the middle of a frame? If that's the case, it would be pretty bad behavior.

No. An in-progress transmission will always run to its end. If the transmission succeeded, the frame is considered as transmitted "OK" and the abort request is ignored. If the in-progress transmissions fails for some reason (arbitration lost, error frame, ...) the TX MB is rescheduled, but since it is now marked for abortion _before_ starting transmission, the transmission will never begin and the frame is aborted.

> Looks like the current approach is safe. Init mode stops further transmissions and the queues are cleared afterwards.

Right.

Reply:
Thanks for the clarification. The handing should be fine indeed.